File: code/Python/Chapter4/4-1/stack.py
class Stack:
    def __init__(self):
        self.__data = []
    
    def is_empty(self):
        return self.__data == []
    
    def __len__(self):
        return len(self.__data)
    
    def push(self, elem):
        self.__data.append(elem)
    
    def pop(self):
        return self.__data.pop()
    
    def peek(self):
        return self.__data[-1]

File: code/Java/Chapter4/4-1/Stack.java
import java.util.ArrayList;

public class Stack<T> {
    private ArrayList<T> data;

    public Stack() {
        data = new ArrayList<T>();
    }

    public int size() {
        return data.size();
    }

    public boolean isEmpty() {
        return data.size() == 0;
    }

    public void clear() {
        data.clear();
    }

    public void push(T elem) {
        data.add(elem);
    }

    public T pop() {
        if (data.size() == 0) {
            return null;
        }
        return data.remove(data.size() - 1);
    }

    public T peek() {
        if (data.size() == 0) {
            return null;
        }
        return data.get(data.size() - 1);
    }
}
